### PR TITLE
[RFR] Parser strategies

### PR DESCRIPTION
**Pull request by djhi**
[Trello card](https://trello.com/c/IWr2WOQy/7-implementer-une-notion-de-strategie-de-correction-fix-github-update)

- [x] Refactored parser to allow different strategies
- [x] Refactored lamba so that it can handle an array of comment parsed from github event
- [x] Tests

### CODE REVIEW

File: src/parser/index.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        const event = getGitHubEventHeader(request.headers);
+
+        if (!event) {
+            logger.debug('no event in github payload');
+            return [];
+        }
+
+        logger.debug('event in github payload', event);
+
+        if (event === 'ping') return [];
```
Comment (Kmaschta) [inline src/parser/index.js]: You can use `ping: null,` in `parsers` to ease the readability.

File: src/parser/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+        const parser = parsers[event];
+        if (!parser) return [];
```
Comment (Kmaschta) [inline src/parser/index.js]: Add a log / warning "No parser found"?

File: src/parser/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
+import { assert } from 'chai';
+import parseFactory from './';
+
+describe.only('Parser', () => {
```
Comment (Kmaschta) [inline src/parser/index.spec.js]: Only spotted

File: src/parser/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import getGitHubEventHeader from './getGitHubEventHeader';
+import parsePullRequestReviewFactory from './parsePullRequestReview';
+import parsePullRequestReviewComment from './parsePullRequestReviewComment';
+
+export default (client, logger) => {
+    const parsers = {
+        ping: null,
+        pull_request_review_comment: parsePullRequestReviewComment,
+        pull_request_review: parsePullRequestReviewFactory(client),
```
Comment (Kmaschta) [inline src/parser/index.js]: Do you need to call the factory each time?
I would prefer `parsers[event](client)` and use it in factory only if needed.

Comment (djhi) [inline src/parser/index.js]: but not every parsers need the client. Why do you want this ?

**Comment by djhi** [comment]
@Kmaschta 
> But is the parsedContents instead of parsedContent still needed ?
I don't get it

File: src/index.js [hunk 5]
```diff
@@ -1,58 +1,66 @@
 /* global config */
 import co from 'co';
+import defer from 'co-defer';
```
Comment (Kmaschta) [inline src/index.js]: Is this new dependency really used?

File: src/index.js [hunk 6]
```diff
@@ -1,58 +1,66 @@
...
+            owner: parsedContents[0].repository.user,
+            name: parsedContents[0].repository.name,
         },
     });
 
     const fixer = fixerFactory(git, logger);
-    const fixedContent = yield fixer.fix(parsedContent);
-    logger.debug('Content fixed', { fixedContent });
+    const fixedContents = [];
+    for (let i = 0; i < parsedContents.length; i++) {
```
Comment (Kmaschta) [inline src/index.js]: `for (const parsedContent of parsedContents)` ?

File: src/git/commands/checkout.js [hunk 7]
```diff
@@ -6,12 +6,14 @@ export default (references, commits) => function* (ref, force = false) {
     // If no: continue
 
     const commitSha = yield references.get(ref);
+    console.log('checkout', { ref, commitSha, force });
```
Comment (Kmaschta) [inline src/git/commands/checkout.js]: console.log spotted

File: src/git/commands/checkout.js [hunk 8]
```diff
@@ -6,12 +6,14 @@ export default (references, commits) => function* (ref, force = false) {
...
 
     if (!commitSha) {
         throw new CommandError(`Reference ${ref} does not exist`);
     }
 
     const commit = yield commits.get(commitSha);
+    console.log('checkout', { commit });
```
Comment (Kmaschta) [inline src/git/commands/checkout.js]: idem

**Pull request merged**